src/OpenTelemetry.Extensions.Owin/DiagnosticsMiddleware.cs

```diff
@@ -0,0 +1,158 @@
+    public sealed class DiagnosticsMiddleware : OwinMiddleware
+    {
+        private const string ActivityName = "OpenTelemetry.Extensions.Owin.HttpRequestIn";
+        private const string ActivityStartKey = ActivityName + ".Start";
+        private const string ActivityStopKey = ActivityName + ".Stop";
+
+        private readonly DiagnosticListener diagnosticListener = new DiagnosticListener("OpenTelemetry.Extensions.Owin");
+        private readonly Context context = new Context();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DiagnosticsMiddleware"/> class.
+        /// </summary>
+        /// <param name="next">An optional pointer to the next component</param>
+        public DiagnosticsMiddleware(OwinMiddleware next)
+            : base(next)
+        {
+        }
+
+        /// <inheritdoc />
+        public override async Task Invoke(IOwinContext owinContext)
+        {
+            try
+            {
+                this.BeginRequest(owinContext);
+                await this.Next.Invoke(owinContext).ConfigureAwait(false);
+                this.RequestEnd(owinContext, null);
+            }
+            catch (Exception ex)
+            {
+                this.RequestEnd(owinContext, ex);
+            }
+        }
+
+        // Based on https://github.com/dotnet/aspnetcore/blob/v5.0.0-rc.2.20475.17/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L37
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void BeginRequest(IOwinContext owinContext)
+        {
+            if (OwinExtensionsEventSource.Log.IsEnabled())
+            {
+                this.context.EventLogEnabled = true;
+            }
+
+            if (this.diagnosticListener.IsEnabled() && this.diagnosticListener.IsEnabled(ActivityName, owinContext))
+            {
+                this.context.Activity = this.StartActivity(owinContext, out var hasDiagnosticListener);
+                this.context.HasDiagnosticListener = hasDiagnosticListener;
+            }
+        }
+
+        // Based on https://github.com/dotnet/aspnetcore/blob/v5.0.0-rc.2.20475.17/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L89
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void RequestEnd(IOwinContext owinContext, Exception exception)
+        {
+            var activity = this.context.Activity;
+            // Always stop activity if it was started
+            if (activity != null)
+            {
+                this.StopActivity(owinContext, activity, this.context.HasDiagnosticListener);
+            }
+
+            if (this.context.EventLogEnabled && exception != null)
+            {
+                // Non-inline
+                OwinExtensionsEventSource.Log.UnhandledException();
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private Activity StartActivity(IOwinContext owinContext, out bool hasDiagnosticListener)
```

Comment:
The recommendation is to use `ActivitySource` based instrumentation for any new instrumentations. DiagnosticListener based ones are "legacy".
This instructions are being documented here (WIP): 
https://github.com/open-telemetry/opentelemetry-dotnet/pull/1467

I can help with replacing the DS -> ActivitySource ones.

Reply:
Thanks @cijothomas, I'll replace `DiagnosticListener` with `ActivitySource`.